### DISCUSSION TITLE
UMP and IAP Europe Transparency & Concent Framework

### DISCUSSION THREAD

**Discussion by merbin2012** [EMI-INDO/emi-indo-cordova-plugin-admob#31] [Q&A]
Can you please help me to write the proper code for the following
User Messaging Platform (iOS & Android)
IAB Europe Transparency & Consent Framework

Currently, I am using admob plus & its consent plugin. There I am using the following code

```
			if (cordova.platformId === 'ios') {
				const status = await consent.trackingAuthorizationStatus()
				/*
				  trackingAuthorizationStatus:
				  0 = notDetermined
				  1 = restricted
				  2 = denied
				  3 = authorized
				*/
				const statusNew = await consent.requestTrackingAuthorization()
			}

			/* const consentStatus = await consent.getConsentStatus()
			if (consentStatus === consent.ConsentStatus.Required)
			{
			  if(_locViewCount==1)
				  return;
	  
			} */

			//if (consentStatus === consent.ConsentStatus.Required) {

			//}
			await consent.requestInfoUpdate()
			await consent.loadAndShowIfRequired()

			 //ad load and init events code are written 
```

And, my results are 
![IDFA](https://github.com/user-attachments/assets/4a3785e9-fc30-455f-a8f4-e54da6f728c6)
![GDPR](https://github.com/user-attachments/assets/c9e36ded-d237-421b-93db-0ae30fbe2d1c)

Already I have many confusions in admob plus consent plugin, here I don't have any idea to implement it. Also, here you are using two plugin instead of admob plus using one plugin. Can you please give clear examples and explanation.

**Note:** In android we can force users to accept the consent and show the ad, if no we can close the app. But in iOS, it is not allowed.

**Comment by EMI-INDO** [comment]
Does this not work, or do you need a different logic?

```
cordova.plugins.emiAdmobPlugin.requestIDFA(); // requestTrackingAuthorization
```


https://github.com/EMI-INDO/emi-indo-cordova-plugin-admob/blob/main/example/www/js/deviceready.js



 // SDK EVENT Initialization
    // Optional
```
    document.addEventListener('on.sdkInitialization', (data) => {
    
        
        if (isPlatformIOS){
            
            log.value += ("\n On Authorization Status: " + data.attStatus);
        }

    });
```

**Comment by merbin2012** [comment]
It seems working for iOS but what to do for Android consent form?
What is the usage of this
User Messaging Platform (iOS & Android)
IAB Europe Transparency & Consent Framework

**Comment by EMI-INDO** [comment]
If the current or previous ATT is not approved or Not Determined, the UMP will be displayed automatically,  

If ATT is currently or previously approved the UMP is not displayed, but you can display the UMP with `cordova.plugins.emiAdmobPlugin.showPrivacyOptionsForm();`

- even by force

`cordova.plugins.emiAdmobPlugin.forceDisplayPrivacyForm()`



https://developers.google.com/admob/ios/privacy/idfa


While testing, remember that the IDFA ATT dialog only appears a single time since [requestTrackingAuthorization](https://developer.apple.com/documentation/apptrackingtransparency/attrackingmanager/3547037-requesttrackingauthorization) is a one-time request. The UMP SDK only has a form available to load if the authorization status is [ATTrackingManagerAuthorizationStatusNotDetermined](https://developer.apple.com/documentation/apptrackingtransparency/attrackingmanagerauthorizationstatus/attrackingmanagerauthorizationstatusnotdetermined?language=objc).

To make the alert appear a second time, you must uninstall and reinstall your app on your test device.

**Comment by EMI-INDO** [comment]
ATT decisions cannot be changed, except for uninstalling and reinstalling the app.

The way to change the ATT decision may be possible with UMP, but I don't know for sure.

- Example:
In the application by setting up a button on click or other logic and executing the function

`cordova.plugins.emiAdmobPlugin.showPrivacyOptionsForm();`

even by force
- Note Only IOS
`cordova.plugins.emiAdmobPlugin.forceDisplayPrivacyForm()`

you can test it try not approve ATT

`data.attStatus`




continue to do just say on button click


`cordova.plugins.emiAdmobPlugin.showPrivacyOptionsForm();`

even by force
- Note Only IOS
`cordova.plugins.emiAdmobPlugin.forceDisplayPrivacyForm()`

`data.attStatus`

**Comment by EMI-INDO** [comment]
### From gpt chat

In iOS, there are two types of permissions related to tracking and user privacy: UMP (User Messaging Platform) and ATT (App Tracking Transparency). Although both are related to tracking user data, they have different purposes and ways of implementation:

### 1. ATT (App Tracking Transparency)
Purpose: ATT is used to request users' permission before apps can track their activity across other apps or websites (e.g., tracking-based ads). This is regulated by Apple to maintain user privacy.
Application: If an app wants to do tracking across apps or sites (for example, using IDFA - Identifier for Advertisers), you must call the ATT permission dialog. The user will be given two choices: “Request Permission” or ‘Disallow’.
How it works: If the user chooses not to allow tracking, your app cannot access IDFA and cannot perform cross-app or cross-site tracking.

### 2. UMP (User Messaging Platform)
Purpose: UMP is used to display permissions related to advertising or data collection relevant to app usage, and it was deployed after the launch of ATT. UMP allows you to display a dialog related to personal data collection or to provide information about the app's privacy policy.
Implementation: UMPs are often used to display further dialogs after ATT permissions have been granted, which focus more on the app's privacy policy or the use of data in the app.
Sequence Performed in Code
ATT is executed first. Since ATT is a requirement for tracking user activity outside your app, you must first ask for ATT permission. If the user refuses ATT, then you cannot proceed to track or use IDFA.
Once ATT permission is granted, UMP can be run. If your app requires the display of additional information or requests permission regarding in-app data collection, then you can proceed with using the UMP after the ATT is approved by the user.


### Implementation Sequence in Code:
Check the ATT permission first and display the dialog if needed.
If the ATT is approved, proceed with the UMP permission or dialog if required.


### Conclusion
- ATT Approved: Personalized ads can be displayed.
- ATT Rejected: Ads will be non-personalized.
- UMP Not Implemented: Will not affect ad display as long as ATT is implemented, but it may violate privacy policies in certain regions (such as GDPR in Europe or CCPA in California), which require you to provide additional setting options to users.

If your app operates in regions like Europe, it is recommended to stick with UMP or similar tools to comply with local privacy regulations.

**Comment by merbin2012** [comment]
So, no need to install separate plugin like admob plus consent right? 
Also, except iOS (cordova.plugins.emiAdmobPlugin.requestIDFA();), android no need any special code for this concept and the form will be display if required automatically while request the ads right?

**Comment by EMI-INDO** [comment]
- Yes you are right
- Because startGoogleMobileAdsSDK is always inside the UMP block.


### On IOS method 

cordova.plugins.emiAdmobPlugin.initialize({...})

- objective-c https://github.com/EMI-INDO/emi-indo-cordova-plugin-admob/blob/main/src/ios/emiAdmobPlugin.m
- Line code start 85-87

```
 if (UMPConsentInformation.sharedInstance.canRequestAds) {
    [self startGoogleMobileAdsSDK];
  }
```

As explained here https://developers.google.com/admob/ios/privacy#request-ads

### request ads

It is possible consent has been obtained in the previous session. As a latency best practice, we recommend not waiting for the callback to complete so you can start loading ads as soon as possible after your app launches.


### Likewise in android
- Because initializeMobileAdsSdk() is always inside the UMP block.

### On Android method 

cordova.plugins.emiAdmobPlugin.initialize({...})

- kotlin https://github.com/EMI-INDO/emi-indo-cordova-plugin-admob/blob/main/src/android/emiAdmobPlugin.kt
- Line code start 283-285

```
if (consentInformation?.canRequestAds()!!) {
                        initializeMobileAdsSdk()
                    }
```

**Comment by EMI-INDO** [comment]
If something goes wrong during the consent collection process, it doesn't matter because it has been handled by

```
{ formError: FormError ->
       if (consentInformation!!.canRequestAds()) {

               initializeMobileAdsSdk()
       }
                             
 })
```

**Comment by merbin2012** [comment]
Thank you.
Also, can you suggest me to use ad manager or admob ( isUsingAdManagerRequest).
Long months ago, I was tried FAN mediation with admob but I think it is not given higher earnings. What is your suggestion? Please explain your experience in mediation.

**Comment by EMI-INDO** [comment]
Based on my experience, mediation may be necessary only in certain applications, 
For example if your app users are mostly from high eCPM countries, which allows you to specify an eCPM of say $10 or higher, if the app users are mostly from low eCPM as such mediation is not useful.

And one more thing in high eCPM countries adding dependencies is not enough, consent and GDPR must also be required, 
for example
### Unity Ads

```
dependencies {
    implementation("com.google.android.gms:play-services-ads:23.5.0")
    implementation("com.unity3d.ads:unity-ads:4.12.4")
    implementation("com.google.ads.mediation:unity:4.12.4.0")
}



import com.unity3d.ads.metadata.MetaData;

MetaData gdprMetaData = new MetaData(this);
gdprMetaData.set("gdpr.consent", true);
gdprMetaData.commit();
```



If someone needs consent on mediation I can add it, but 
until now it's not there.

**Comment by merbin2012** [comment]
Okay. Thank you. 

Note Check this https://github.com/EMI-INDO/emi-indo-cordova-plugin-admob/discussions/29#discussioncomment-11390497
